nengo/synapses.py

```diff
@@ -221,123 +197,123 @@ def evaluate(self, frequencies):
         y = np.polyval(self.num, w) / np.polyval(self.den, w)
         return y
 
-    def make_step(self, shape_in, shape_out, dt, rng, y0=None,
-                  dtype=np.float64, method='zoh'):
-        """Returns a `.Step` instance that implements the linear filter."""
+    def _get_ss(self, dt, method='zoh'):
+        A, B, C, D = tf2ss(self.num, self.den)
+
+        # discretize (if len(A) == 0, filter is stateless and already discrete)
+        if self.analog and len(A) > 0:
+            A, B, C, D, _ = cont2discrete((A, B, C, D), dt, method=method)
+
+        return A, B, C, D
+
+    def allocate(self, shape_in, shape_out, dt,
+                 y0=None, dtype=np.float64, method='zoh'):
         assert shape_in == shape_out
 
-        num, den = self.num, self.den
-        if self.analog:
-            num, den, _ = cont2discrete((num, den), dt, method=method)
-            num = num.flatten()
-
-        if den[0] != 1.:
-            raise ValidationError("First element of the denominator must be 1",
-                                  attr='den', obj=self)
-        num = num[1:] if num[0] == 0 else num
-        den = den[1:]  # drop first element (equal to 1)
-        num, den = num.astype(dtype), den.astype(dtype)
-
-        output = np.zeros(shape_out, dtype=dtype)
-        if len(num) == 1 and len(den) == 0:
-            return LinearFilter.NoDen(num, den, output)
-        elif len(num) == 1 and len(den) == 1:
-            return LinearFilter.Simple(num, den, output, y0=y0)
-        return LinearFilter.General(num, den, output, y0=y0)
-
-    @staticmethod
-    def _make_zero_step(shape_in, shape_out, dt, rng, y0=None,
-                        dtype=np.float64):
-        output = np.zeros(shape_out, dtype=dtype)
-        if y0 is not None:
-            output[:] = y0
-
-        return LinearFilter.NoDen(np.array([1.]), np.array([]), output)
+        A, B, C, D = self._get_ss(dt, method=method)
+        X = np.zeros((A.shape[0],) + shape_out, dtype=dtype)
+
+        if y0 is not None and len(X) > 0:
+            y0 = np.array(y0, copy=False, ndmin=1)
+            X[:] = np.linalg.solve(np.eye(len(A)) - A, B * y0)
+
+        return dict(X=X)
+
+    def make_step(self, shape_in, shape_out, dt, rng, state=None,
+                  method='zoh'):
+        """Returns a `.Step` instance that implements the linear filter."""
+        assert shape_in == shape_out
+        assert state is not None
+
+        A, B, C, D = self._get_ss(dt, method=method)
```

Comment:
Is this being called twice in a row? One by the allocation and again by `make_step`?

Reply:
Yes. It didn't seem worth it to cache because it's quite cheap.

Reply:
One potential issue I see later is if `method` is refactored or exposed in some way. Currently we're trusting that it's called with the exact same signature in each place, which opens up some surface area for things to go wrong. Is there a way to avoid this without caching?

Reply:
Yeah, I was wondering about having `method` as a LinearFilter class attribute.